src/coreclr/jit/sideeffects.h

```diff
@@ -128,8 +128,12 @@ class AliasSet final
 
             if ((m_flags & ALIAS_WRITES_LCL_VAR) != 0)
             {
+                // Stores to 'lvLiveInOutOfHndlr' locals cannot be reordered with
+                // exception-throwing nodes so we conservatively consider them
+                // globally visible.
+
                 LclVarDsc* const varDsc = m_compiler->lvaGetDesc(LclNum());
-                return varDsc->IsAlwaysAliveInMemory();
+                return varDsc->lvLiveInOutOfHndlr != 0;
```

Comment:
Worked with @SingleAccretion on this fix.

The premise is that `IsAlwaysAliveInMemory` checks both `lvLiveInOutOfHndlr` and `lvSpillAtSingleDef`. The latter is only set in LSRA and shouldn't have any significance with regards to side effect checking since such spills can't interfere.

This solves the inconsistency between the check in lowering and the assert in codegen.